README.md
# CueStep Fader

## Disclaimer
__ATTENTION!__
You are using this plugin at your own risk!
Even though there have been no critical bugs reported until now, the creator of this plugin under no circumstances assumes responsibility for damaged, destroyed or in any way negatively influenced show files or damaged or destroyed content or parts of a show file.
Please always do a backup on an external USB-Stick BEFORE running this plugin.
DO NOT RUN THIS PLUGIN IN ANY SHOW-CRITICAL SITUATIONS!


## Table of contents
* [What is CueStep Fader?](#what-is-cuestep-fader?)
* [How to use it](#how-to-use-it)
* [Installation guide](#installation-guide)
* [Developer infos](#developer-infos)
* [Technology](#technology)


## What is CueStep Fader?
CueStep Fader is a Lua plugin for GrandMA2 lighting consoles and its goal is to make cue-step fader usable for everyone!
With CSF you can trigger specific cues in a sequence depending on the faderposition. For example if you have a 10 channel sunstrip and want to control how many lamps are glowing by moving the fader up and down. And the best: It doesn't require any in-depth knowlege of the console and sets up everything in seconds and nearly completely automatically!


## How to use it
CueStep Fader is very simple to use. Once it is installed (see installation guide below), you can either click on the plugin in the plugin pool window or type 'Plugin "CueStep Fader"' into the command line, to start it.
1. Enter the amount of steps you need, as an integer. Mind, that the maximum number is steps is currently limited to 100 for technical reasons.
2. Enter your preferred executor number by using either the format PAGE.EXECUTOR or EXECUTOR (for using the current page).
3. Enter a name for your CSF. The name must not contain dots but _can_ be empty - however this is not recommended in terms of later identification.


## Installation guide
Please follow these Steps for the installation of CSF:
1. Move the required files to the right directory.<br> 
   1.1 For onPC: Copy the folder named 'CueStep Fader' into: C:\Program Data\MA Lighting Technologies\grandma\gma2_V_X\plugins.<br>
   1.2 For Console: Copy the folder 'CueStep Fader' to a USB-Stick at: „STICK NAME"\gma2\plugins.
   _Important(for both cases): The folder HAS to be labeled as 'CueStep Fader'. Otherwise the Plugin won't run!_
2. Open GrandMA2 onPC / start console.
3. Open a PLUGIN-Pool. You can find it under „System".
4. Right-click on a free plugin-field.
5. Click at "import", navigate to 'CueStep Fader/Plugin' and choose „CSP2.xml"
6. Close the window.

## Developer infos
### General informations
Plugin versions are based on "Semantic Versioning 2.0.0". Visit [semver.org](https://semver.org) for more informations.

### Automated testings
To test the plugin automatically with the Test.lua script, import the plugin as described above. But instead of running the plugin directly type<br>
    ```LUA 'CSF_main(true)'```<br>
into the commandline, as it runs CSF in test mode.<br>
A Test.log file is created under /ProgramData/MA Lighting Technologies/grandma/gma2_V_X.X.<br>
_Note, that the log file is empty until onPC.exe is closed._

## Technology
* Lua 
* XML


Changelog.md
# V 0.27.0
## Added features:
--

## Fixed bugs:
1. Test.log file can only be viewed when the GrandMA Software is shut down.

## Enhancements:
### API CHANGES:
1. All functions that are about managing fixture types and there files are now outsourced to the csfixtype module for clarity reasons.
2. Versioning of the 'csfixtype' module changed to match semantic versioning (2.0.0). csfixtype is now on version 1.0.0.

---

# V 0.26.1
## Added features:
--

## Fixed bugs:
1. When multiple CSFs are created, only the first one works correctly. All following CSFs are triggered erroneously by the first one or don't work at all due to wrong dmx addresses assigned to the dmx remotes.
2. When number of steps is too high, not alle steps can be triggered by the CSF. (Limit for steps is now 100!)
3. DMX remotes don't work when the CSF name contains a dot. (A dot is now an illegal character for the name input.)

## Enhancements:
--

---

# V 0.26.0
## Added features:
1. More then one universe can be used for CSFs if necessary.

## Fixed bugs:
--

## Enhancements:
1. CSF is now executable, even if Test.lua file is absent.

---

# V 0.25.0
## Added features:
1. Can be tested automatically by using an added test mode.

## Fixed bugs:
1. Console can't import the necessary modules due to a wrong 'package.path configuration.'

## Enhancements:
--

---

# V 0.24.2
## Added features:
--

## Fixed bugs:
--

## Enhancements:
1. Enhanced precision by created DMX-Profiles.
2. In case of an invalid user input, an error message is displayed.

---

# V 0.24.1
## Added features:
--

## Fixed bugs:
1. Invalid Executornumber 'PAGE'.00 is chosen when all Executors on the chosen page are taken.
    FIXED.
2. Invalid Executornumber 'PAGE'.0.01 is chosen when all Executors on the chosen page are taken
   and the preferred executor number were entered without an page (e.g. "10" instead of "1.10")
   FIXED.

## Enhancements:
1. CSContainer can be moved to any executor.
2. Using CSF names twice is prevented by requesting user to enter the CSF name
   until the chosen name is not already in use.
3. The CSFader now gets labeled according to the scheme "[CSF Name] CSF".

---

# V 0.24.0
## Added features:
1. CSContainer is now assigned to an executor, to enable executor options.
    - The system variable $CSC_[CSF NAME] is added which contains the executor number of the CSContainer. 
      If CSContainer is moved, it has to be updated.

## Fixed bugs:
--

## Enhancements:
1. CSFader is now swop protected.
2. "React To Master" is disabled for CSFixtures

---

# V 0.23.7
## Added features:
--

## Fixed bugs:
--

## Enhancements:
1. Performance enhancements through code optimizations.

---

# V 0.23.6
## Added features:
--

## Fixed bugs:
1. When running CSF more then one time, CSLayer gets recreated and overwritten every time.
    Fixed.
    Changings for the User:
    1.1. The second CSFixture is appended to the existing Layer and the existing Universe.
    1.2. Added GrandMA2 show variables $CSF_ADDR_CACHE and $CSF_UNI to enable patching and using multiple CSFixtures.
         These variables shall not be changed by the user.
2. When "1.10" is entered as executor number, it gets misinterpreted as "1.1"
    Fixed.
3. Finding an alternative executor fails sometimes and leads to a plugin crash.
    Fixed.


## Enhancements:
1. DMX Profiles can now be associated with the mode and the fixture version they were created with. 
   Schema is "CSF_Profile [INDEX]/[STEPS]_[VERSION]" (without []).
2. Performace enhancements through code optimizations.


Test.lua
-- This file is for automatically testing and logging CueStep Fader.
-- 0.1.2

Test = {} -- module table to be returned
local cmd = function(syntax, ...) gma.cmd(syntax:format(...)) end
local LOGFILE = io.open("Test.log", "w+")
io.output(LOGFILE)
io.write(
    '-- Values to be tested for every input: --\n',
    'emtpy string\n',
    '0\n',
    '1\n',
    '10\n',
    '100\n',
    '200\n',
    '1.5\n',
    'a\n',
    'bla \\ bla\n',
    '"test" tada\n\n',
    '---------------------------------------\n\n'
)

local function log(varname, var, ...)
    --[[
        Writes logging / debugging infos to LOGFILE.
    ]]
    if varname ~= nil then
        io.write(varname..": "..(var or "").."| ", ... or "", "\n")
    else
        io.write(..., "\n")
    end
    io.flush()
end

local function all_tested()
    --[[
        Checks if all listed inputs have been tested and returns true if this is the case.
        Otherwise it returns false.
    ]]
    log("steps.num_of_tested", Test.steps.num_of_tested)
    log("execnum.num_of_tested", Test.execnum.num_of_tested)
    log("name.num_of_tested", Test.name.num_of_tested)
    if Test.steps.num_of_tested < #Test.steps.inputs then
        return false
    end
    if Test.execnum.num_of_tested < #Test.execnum.inputs then
        return false
    end
    if Test.name.num_of_tested < #Test.name.inputs then
        return false
    end
    return true
end

Test.test = coroutine.create(function()
    --[[
        The main function for the test. It runs CSF repeatedly (with param. 'testmode' set to true)
        until all inputs listed in Test.lua are tested.
    ]]
    local run = 1
    while not all_tested() do
        log("Run", run, 'Not all tested. Next one.')
        cmd('LUA "CSF_main(true)"')
        coroutine.yield() -- wait for the current run of CSF to finish.
        log(nil, nil, "\n Resume yielding loop.")
        run = run+1
    end
    log(nil, nil, "Test completed.")
    LOGFILE:close()
end)

Inputs = {} -- class / metatable
function Inputs:new(name) -- init instance of Inputs
    local o = {}
    setmetatable(o, self)
    self.__index = self
    o.name = name
    o.num_of_tested = 0
    o.inputs = {
        { val = '', tested = false },
        { val = '0', tested = false },
        { val = '1', tested = false },
        { val = '10', tested = false },
        { val = '100', tested = false },
        { val = '200', tested = false },
        { val = '1.5', tested = false },
        { val = 'a', tested = false },
        { val = 'bla \\ bla', tested = false },
        { val = '\"test\" tada', tested = false },
    }

    function o:get_new_val()
    --[[
        Returns the first untested value.
        If all values have already been tested, it returns "2" as it's an valid input in anycase.
    ]]
        for _, input in ipairs(self.inputs) do
            if not input.tested then
                input.tested = true
                self.num_of_tested = self.num_of_tested + 1
                log("new_val", input.val, 'for '..self.name)
                return input.val
            end
        end
        log(nil, nil,"All values for "..self.name.." tested. Return '2'")
        return "2"
    end
    return o
end

Test.steps = Inputs:new("steps")
Test.execnum = Inputs:new("execnum")
Test.name = Inputs:new("name")

return Test

Plugin/csfixtype.lua
csfixtype = {}
csfixtype.VERSION = {major = 1, minor = 0, patch = 0, str = "1.0.0"}

local pooltools = require 'pooltools'
local x_cache

local function get_files(path, op_sys)
    --[[
        Searches for existing fixture type files and import their names into the table file_names,
        which is returned at the end.
        Parameters: path (str), op_sys(str)
        Return value: file_names (table)
    ]]
    local file_names = {}
    local list_cmd = ""
    if op_sys == "WINDOWS" then
        list_cmd = 'dir /B "'..path..'" | findstr /i "csfixtype.*.xml"'
    else
        list_cmd = 'ls "'..path..'" | grep "csfixtype.*.xml"'
    end
    local file_list = io.popen(list_cmd)
    local files = file_list:lines()
    for file in files do
        table.insert(file_names, file)
    end
    return file_names
end

local function name_to_vers(name)
    --[[
        Takes the filename as a parameter (str) and returns a table containing the major, minor and patch
        version.
    ]]
    local version_pos = name:find("v") + 1
    local version_str = name:sub(version_pos)
    local get_subversion = version_str:gmatch("%d")
    local version = {
        major = tonumber(get_subversion()),
        minor = tonumber(get_subversion()),
        patch = tonumber(get_subversion())
    }
    return version
end

local function is_deprecated(fname)
    --[[
        Returns true if the filename passed as a parameter (str), indicates a fixture type
        version that is not up to date. Returns false otherwise.
    ]]
    local file_vers = name_to_vers(fname)
    if file_vers.major < csfixtype.VERSION.major then
        return true
    elseif file_vers.minor < csfixtype.VERSION.minor then
        return true
    elseif file_vers.patch < csfixtype.VERSION.patch then
        return true
    end
    return false
end

local function delete_deprecated(path, op_sys)
    --[[
        Checks the library for deprecated CSF fixture types and deletes them.
        Parameter: path (to the library; str), op_sys (str)
        Returns nil.
    ]]
    local deleted_files_count = 0
    local files = get_files(path, op_sys)
    for _, file in pairs(files) do
        if is_deprecated(file) then
            os.remove(path..'/'..file)
            deleted_files_count = deleted_files_count + 1
        end
    end
    gma.echo(deleted_files_count..' deprecated fixturetypes deleted.')
end

local function calc_x(step_size)
    --[[
        Calculates the X coordinates for the DMX profile points.
        Parameters: step_size (number)
        Return value: t (table)
    ]]
    local a = x_cache or 0
    local b = (a*10 + 0.001)/10 -- multiply and divide by 10 to get more precise results!
    local c = a + step_size
    local d = (c*10 + 0.001)/10
    x_cache = c
    local t = {x0 = a, x1 = b, x2 = c, x3 = d}
    return t
end

local function create_fixtype(file, number_of_steps)
    --[[
        Creates a grandMA2 fixture type as an XML file depending on the given parameters.
        Parameters: file (file obj), number_of_steps (number)
        Returns nil.
    ]]
    local profile_index = pooltools.getFreeObj('Profile', 1, number_of_steps+1)
    local step_size = 1 / number_of_steps
    local points = {}
    local vers = table.concat(csfixtype.VERSION, ".")
    io.output(file)
    io.write(string.format(
        '<?xml version="1.0" encoding="utf-8"?>\n'..
        '<MA xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://schemas.malighting.de/grandma2/xml/MA" xsi:schemaLocation="http://schemas.malighting.de/grandma2/xml/MA http://schemas.malighting.de/grandma2/xml/3.7.0/MA.xsd" major_vers="3" minor_vers="7" stream_vers="0">\n'..
	    '\t<Info datetime="2019-11-26T13:18:25" showfile="" />\n'..
	    '\t<FixtureType name="CueStep Fader" mode="%i steps">\n'..
		'\t\t<short_name>CSF</short_name>\n'..
		'\t\t<manufacturer>JonasKochMarkHuber</manufacturer>\n'..
        '\t\t<short_manufacturer>JKMH</short_manufacturer>\n'..
        '\t\t<revision date="10.09.2020" text="Version %s" generator_software_name="CueStep Fader" generator_software_version="%s" />\n'..
		'\t\t<Profiles>', number_of_steps, vers, vers
    ))
    -- Generator start
    io.write(string.format('\n'..
        '\t\t\t<DMX_Profile index="%i" name="CSF_Profile 0/%i_%s" display_spec_index="1">\n'..
        '\t\t\t\t<DMX_Profile_Point index="0" y="1" mode="linear" />\n'..
        '\t\t\t\t<DMX_Profile_Point index="1" x="0.01" y="1" mode="linear" />\n'..
        '\t\t\t\t<DMX_Profile_Point index="2" x="0.0101" mode="linear" />\n'..
        '\t\t\t</DMX_Profile>\n',
    profile_index, number_of_steps, vers))
    x_cache = 0.01
    for i=profile_index+1, profile_index + number_of_steps-1 do
        points = calc_x(step_size)
        io.write(string.format(
            '\t\t\t<DMX_Profile index="%i" name="CSF_Profile %i/%i_%s" display_spec_index="1">\n'..
            '\t\t\t\t<DMX_Profile_Point index="0" x="%.4f" mode="linear" />\n'..
            '\t\t\t\t<DMX_Profile_Point index="1" x="%.4f" y="1" mode="linear" />\n'..
            '\t\t\t\t<DMX_Profile_Point index="2" x="%.4f" y="1" mode="linear" />\n'..
            '\t\t\t\t<DMX_Profile_Point index="3" x="%.4f" mode="linear" />\n'..
            '\t\t\t</DMX_Profile>\n',
        i, i-profile_index, number_of_steps, vers, points.x0, points.x1, points.x2, points.x3))
    end
    points = calc_x(step_size-0.01)
    io.write(string.format(
        '\t\t\t<DMX_Profile index="%i" name="CSF_Profile %i/%i_%s" display_spec_index="1">\n'..
        '\t\t\t\t<DMX_Profile_Point index="0" x="%.4f" mode="linear" />\n'..
        '\t\t\t\t<DMX_Profile_Point index="1" x="%.4f" y="1" mode="linear" />\n'..
        '\t\t\t\t<DMX_Profile_Point index="2" x="1" y="1" mode="linear" />\n'..
        '\t\t\t</DMX_Profile>\n',
    profile_index + number_of_steps, number_of_steps, number_of_steps, vers, points.x0, points.x1))
    -- Generator end
    io.write(
        '\t\t</Profiles>\n'..
		'\t\t<Modules index="0">\n'..
		'\t\t\t<Module index="0" class="None" beamtype="Wash" beam_angle="35" beam_intensity="10000">\n'..
		'\t\t\t\t<ChannelType index="0" attribute="DIM" feature="DIMMER" preset="DIMMER">\n'..
		'\t\t\t\t\t<ChannelFunction index="0" from="0" to="100" min_dmx_24="0" max_dmx_24="16777215" physfrom="0" physto="1" subattribute="DIM" subattribute_user_name="Dim" attribute="DIM" attribute_user_name="Dim" feature="DIMMER" feature_user_name="Dimmer" preset="DIMMER" preset_user_name="Dimmer" />\n'..
		'\t\t\t\t</ChannelType>\n'
    )
    -- Generator start
    local j = profile_index
    for i=1, number_of_steps+1 do
        io.write(string.format(
            '\t\t\t\t<ChannelType index="%i" attribute="UNKNOWN" feature="CONTROL" preset="CONTROL" coarse="%i" default="254.999" fade_path="%i" react_to_dim="1">\n'..
			'\t\t\t\t\t<ChannelFunction index="0" from="0" to="100" min_dmx_24="0" max_dmx_24="16777215" subattribute="UNKNOWN" subattribute_user_name="Unknown" attribute="UNKNOWN" attribute_user_name="Dummy" feature="CONTROL" feature_user_name="Control" preset="CONTROL" preset_user_name="Control" />\n'..
			'\t\t\t\t</ChannelType>\n',
        i, i, j))
        j = j + 1
    end
    -- Generator end
    io.write(
        '\t\t\t</Module>\n'..
        '\t\t</Modules>\n'..
        '\t\t<Instances index="1">\n'..
        '\t\t\t<Instance index="0" module_index="0" />\n'..
        '\t\t</Instances>\n'..
        '\t\t<Wheels index="2" />\n'..
        '\t\t<VirtualFunctionBlocks index="3" />\n'..
        '\t\t<FixtureMacroCollect index="5" />\n'..
        '\t\t<Body />\n'..
        '\t\t<AutoPresets index="4" />\n'..
        '\t\t<RdmNotifications index="6">\n'..
        '\t\t\t<RdmNotification index="0" Type="Absent" Subcategory="RDM Warning" />\n'..
        '\t\t</RdmNotifications>\n'..
        '\t</FixtureType>\n'.. 
        '</MA>'
    )
    file:close()
    gma.echo('CSFixtype created.')
end

function csfixtype.manage(path, num_of_steps, op_sys)
    --[[
        Triggers the deletion of deprecated fixture type files and creates a new fixture type file
        in case there doesn't already exist a matching file.
    ]]
    delete_deprecated(path, op_sys)
    local fname = string.format('csfixtype-%i--v%s.xml', num_of_steps, csfixtype.VERSION.str)
    local file = io.open(path..'/'..fname, 'r')
    if file then
        file:close()
    else
        file = io.open(path..'/'..fname, 'w')
        create_fixtype(file, num_of_steps)
    end
end

return csfixtype

Plugin/pooltools.lua
-- V 2.0.0

local pooltools = {}
local gethandle = gma.show.getobj.handle

function pooltools.getFreeObj(target, start, amount)
    --[[
        Searches for 'amount' consecutive free pool slots of a specific pool type ('target') starting at slot number 'start'.
        Returns the number of the first suitable slot. 'momode' prints the result in the system monitor.
    ]]--
    local free_slots = 0
    while true do
        local firsttarget = target..' '..start
        local handle      = gethandle(firsttarget)
        if not handle then
            for i = start, start+amount-1 do
                local multitarget = target..' '..i
                local multihandle = gethandle(multitarget)
                if not multihandle then
                    free_slots = free_slots + 1
                else
                    free_slots = 0
                    break
                end
                if free_slots >= amount then return start end
            end
        end
        start = start+1
    end
end

function pooltools.getUni(start_uni, start_addr, amount)
    --[[
        Returns the first matching universe while all parameters are optional.
        start_uni: The universe at which the search should start. Is set to 1 if parameter is nil.
        start_addr: The DMX address at which the search at start_uni should start.
                    Is set to 1 if:
                        - parameter is nil
                        - start_uni has not enough consecutive free DMX addresses, so start_uni is counted up.
        amount: The amount of required consecutive free DMX addresses. Is set to 512 if parameter is nil.
    ]]
    local getprop = gma.show.property.get
    start_uni = start_uni or 1
    start_addr = start_addr or 1
    amount = amount or 512
    local end_uni = 255
    local critical_addr = 512 - amount
    for uni=start_uni, end_uni do
        local target = 'DMX '..uni..'.'
        local free_addr = 0
        for addr=start_addr, 512 do
            local handle = gethandle(target..addr)
            local name = getprop(handle, 1)
            if name == "" then
                free_addr = free_addr + 1
            else
                free_addr = 0
            end
            if free_addr >= amount then return uni end
            if addr > critical_addr and free_addr == 0 then break end
        end
        start_addr = 1
    end
end

function pooltools.getAlt(target, layer, index, range) -- range must be a table! E.g. range = {1, 15}
    --[[
        Returns index of the nearest free pool object within "range" depending on the given index.
        If there's no free pool object within "range", it returns 0.
    ]]
    local alts = {0, 0}
    for i=1, 2 do
        local steps = range[i] > index and 1 or -1
        for j=index, range[i], steps do
            local newtarget = target..' '..layer..'.'..j -- "Target x.y"
            if not gethandle(newtarget) then
                alts[i] = j
                break
            end
        end
    end
    local dist1 = math.abs(index - alts[1])
    local dist2 = math.abs(index - alts[2])
    return dist1 < dist2 and alts[1] or alts[2]
end

return pooltools

Plugin/CSF_main.lua
-- ************************************************
-- CREATED AND DEVELOPED BY JONAS KOCH & MARK HUBER
-- ************************************************
--               Version 0.27.0 (Beta)
-- ************************************************

--********************************************************--
-- *************** Initialize Variables ***************** --
--********************************************************--

local getobj = gma.show.getobj
local getvar = gma.show.getvar
local setvar = gma.show.setvar

local OS = getvar('OS')
if OS == "WINDOWS" then -- Configure Lua's native 'package.path' variable to ensure that the machine finds all necessary modules.
    local module_path = getvar('pluginpath')..'/CueStep Fader'
    package.path = package.path..';'..
    module_path..'/Plugin/?.lua;'..
    module_path..'/?.lua'
else
    local module_path = "/media/%s/gma2/plugins/CueStep Fader"
    package.path = package.path..';'..
    module_path:format('sdb')..'/?.lua;'..
    module_path:format('sdb')..'/Plugin/?.lua;'..
    module_path:format('sdb1')..'/?.lua;'..
    module_path:format('sdb1')..'/Plugin/?.lua;'..
    module_path:format('sdb2')..'/?.lua;'..
    module_path:format('sdb2')..'/Plugin/?.lua;'
end

local csfixtype = require 'csfixtype'
local pooltools = require 'pooltools'
local REQ_SUCCESS, Test = pcall(function() return require 'Test' end)
--^ Prevent the plugin from crashing if "Test.lua" is not available as it is not necessary for it to run.

local cmd = function(syntax, ...) gma.cmd(syntax:format(...)) end
local gethandle = function(syntax, ...) return getobj.handle(syntax:format(...)) end

local DMX_ADDR
local LIB_PATH = getvar("PATH")..'/library'
local NUM_OF_STEPS
local TESTMODE = false -- Is set by calling the main function with 'true' as argument.
local UNI
local userinfo = {}

--********************************************************--
-- ********************* START CODE ********************* --
--********************************************************--

local function get_dmx_address()
    --[[
        Determines the DMX universe and startaddress for the CSF.
        Returns both as separate values of type int.
    ]]
    local amount_of_channels
    local uni = getvar('CSF_UNI')
    local addr = tonumber(getvar("CSF_ADDR_CACHE") or 1)
    if uni then
        uni = tonumber(uni)
        amount_of_channels = NUM_OF_STEPS
    end
    uni = pooltools.getUni(uni, nil, amount_of_channels)
    setvar('CSF_UNI', uni)
    return uni, addr
end

local function verify_execnumber(num, heading)
    --[[
        Requests user to specify preferred execnumber. returns that execnumber if it's available resp.
        returns the next free alternative.
    ]]
    local exec
    local page
    repeat
        num = num or (TESTMODE and Test.execnum:get_new_val() or gma.textinput(heading, ""))
        local border1, border2  = num:find('%.')
        if border1 then
            page = tonumber(num:sub(1, border1-1))
            exec = tonumber(num:sub(border2+1))
        else
            page = tonumber(getvar('Faderpage'))
            exec = tonumber(num)
        end
        heading = "Invalid executor number! Try again."
        num = nil
    until page and exec
    exec = pooltools.getAlt('Executor', page, exec, {1, 15})
    if #tostring(exec) == 1 then exec = '0'..exec end
    local execnum = exec ~= '00' and page..'.'..exec or verify_execnumber(1+page..'.1') 
    --recursively calls verify_execnumber with increasing page numbers until a page with any free executors is found.
    if not gethandle('Page %i', page) then
        for i=tonumber(getvar('Faderpage')), page-1 do
            cmd('page +')
        end
    end
    return execnum
end

local function setup_CSFader(csf_seq, exec, name)
    --[[
        Import and patch the CSFixture and store it to the CSFader.
    ]]
    local dmx_addr = DMX_ADDR
    local ftypename = "CueStep Fader "..NUM_OF_STEPS.." steps"
    local fix_id = pooltools.getFreeObj('fixture', 10001, 1)
    cmd('cd EditSetup; cd Layers')
    local layerhandle = getobj.handle('CSLayer')
    if not layerhandle then
        cmd('store "CSLayer"')
        layerhandle = getobj.handle('CSLayer')
    end
    local fix = getobj.amount(layerhandle)+1
    cmd('cd CSLayer')
    if not gethandle('Fixturetype "%s"', ftypename) then
        local fname = string.format('csfixtype-%i--v%s.xml', NUM_OF_STEPS, csfixtype.VERSION.str)
        local ftypeno = pooltools.getFreeObj('fixturetype', 1, 1)
        cmd('Import "%s" At Fixturetype %i /path="%s"', fname, ftypeno, LIB_PATH)
    end
    cmd('store %i; assign FixtureType "%s" at %i', fix, ftypename, fix)
    cmd('Assign %i /name="CSFixture"; assign %i /fixid=%i; assign %i /patch="%i.%i"; assign %i /ReactToMaster="Off"',
        fix, fix, fix_id, fix, UNI, dmx_addr, fix)
    cmd('cd /')
    cmd('Fixture %i at 100; store seq %i "%s CSF"', fix_id, csf_seq, name)
    cmd('assign seq %i at exec %s', csf_seq, exec)
    cmd('assign exec %s /SwopProtect="On"', exec)
    cmd('clearall')
    dmx_addr = dmx_addr + NUM_OF_STEPS+1 -- Footprint is +1, because of the "CSF_OFF" step
    setvar("CSF_ADDR_CACHE", dmx_addr)
end

local function create_CSContainer(seq, exec, name)
    --[[
        Create an executor containing the steps as cues
    ]]
    exec = verify_execnumber(exec:gsub('%.', '.1'))
    userinfo.csc = seq
    cmd('store seq %i Cue 1 thru %i "Step 1"', seq, NUM_OF_STEPS)
    cmd('label seq %i "CSC_%s"', seq, name)
    cmd('assign seq "CSC_%s" at exec %s', name, exec) 
end

local function setup_remotes(name)
    --[[
        Create the desired DMX remotes.
    ]]
    local remote = pooltools.getFreeObj('remote 3.', 1, NUM_OF_STEPS+1)
    local last_remote = remote + NUM_OF_STEPS
    local dmx_addr = DMX_ADDR
    local step = 0
    cmd('store remote 3.%i thru 3.%i', remote, last_remote)
    cmd('assign remote 3.%i thru 3.%i /Type="CMD"', remote, last_remote)
    for i = remote, last_remote do
        if i == remote then
            cmd('assign remote 3.%i /name="CSF Off" /CMD="off executor *.CSC_%s" /DMX="%i.%i"', i, name, UNI, dmx_addr)
        else
            cmd('assign remote 3.%i /name="CSF %i" /CMD="goto executor *.CSC_%s cue %i" /DMX="%i.%i"', i, step, name, step, UNI, dmx_addr)
        end
        dmx_addr = dmx_addr + 1
        step = step + 1
    end
end

local function print_user_infos()
    --[[
        Informs the user if the plugin was run successfully and if so,
        it prints the location of the CSF-Executor and it's corresponding cue container into a message box.
    ]]
    gma.echo([[CueStep Fader successfully accomplished.
            ************************************
            created by Jonas Koch and Mark Huber
            ************************************]])
    gma.feedback("\n You can find \n - The CSF-Executor at "..userinfo.exec..". \n - The 'cue container' at sequence "..userinfo.csc..".")
    if not TESTMODE then
        gma.gui.msgbox("INFO","You can find \n - The CSF-Executor at "..userinfo.exec..". \n - The 'cue container' at sequence "..userinfo.csc..".")
    end
end

local function validate_steps_input(input)
    --[[
        Checks wether the given parameter input is a natural number less then 100.
        Returns boolean.
    ]]
    if type(input) ~= "number" then return false end
    local is_valid = false
    if input == math.floor(input) and input > 0 and input <= 100 then
        is_valid = true
    end
    return is_valid
end

local function validate_name_input(input)
    --[[
        Checks wether the given parameter input is a legal executorname.
        Returns boolean.
    ]]
    if not input then return false end
    local is_valid = false
    local has_dot = input:match("%.")
    if not has_dot and not getobj.handle("executor *.CSC_"..input) then
        is_valid = true
    end
    return is_valid
end

function CSF_main(testmode)
    --[[
        Requests user input for the number of steps and the CSF-name
        and executes all necessary functions in order.
    ]]
    if testmode and REQ_SUCCESS then
        TESTMODE = true
    elseif not REQ_SUCCESS then
        gma.echo('CSF plugin: Test script not found. Thus test mode is not available.')
    end
    local csfname
    local heading = "How many steps?"
    repeat
        NUM_OF_STEPS = TESTMODE and Test.steps:get_new_val() or gma.textinput(heading, "")
        NUM_OF_STEPS = tonumber(NUM_OF_STEPS)
        heading = "Number of Steps has to be a natural number."
    until validate_steps_input(NUM_OF_STEPS)
    local csfexec = verify_execnumber(nil, "Enter executor number. (e.g 1.1)")
    heading = "Enter a name for the CSF."
    repeat
        csfname = TESTMODE and Test.name:get_new_val() or gma.textinput(heading, "")
        heading = "Name is invalid or already in use."
    until validate_name_input(csfname)
    local csf_seq = pooltools.getFreeObj('sequence', 101, 1)
    local csc_seq = pooltools.getFreeObj('sequence', csf_seq+1, 1)
    UNI, DMX_ADDR = get_dmx_address()
    userinfo.exec = csfexec
    csfixtype.manage(LIB_PATH, NUM_OF_STEPS, OS)
    setup_CSFader(csf_seq, csfexec, csfname)
    create_CSContainer(csc_seq, csfexec, csfname)
    setup_remotes(csfname)
    print_user_infos()
    if TESTMODE then coroutine.resume(Test.test); gma.echo("resumed") end
end

return CSF_main